graph.cpp

```diff
@@ -0,0 +1,194 @@
+
+bool Graph::Contains(int64 nodeid) {
+  return nodemap.find(nodeid) != nodemap.end();
+}
+
+Graph::Node Graph::Node::DeepCopy() {
+  // create a new node
+  auto result = Node();
+  // go through all the incoming and outgoing edges and insert them in the
+  // result. That's all there is to the state of the node, since we are not
+  // storing the id...
+  for(auto v : *outgoing) {
+    result.outgoing->insert(v);
+  }
+  for(auto v : *incoming) {
+    result.incoming->insert(v);
+  }
+  return result;
+}
+
+int64 Graph::AddNode() {
+  // The counter is keeping track of the ids that have been assigned so far.
+  static int64 id_counter = 0;
+  static int64 max_int64 = std::numeric_limits<int64>::max();
+  
+  // The only reason we are doing mod maxint64 and checking that
+  // the id already exists is on the off chance that about
+  // maxint64 addNodes might have already happened on this graph.
+  // Probably not going to happen, but good karma, right?
+  
+  while(Contains(id_counter)) id_counter = (id_counter % max_int64) + 1;
+  nodemap.insert({id_counter, Node()});
+  return id_counter;
+}
+
+int64 Graph::AddNode(int64 id) {
+  if(!Contains(id)) {
+    nodemap.insert({id, Node()});
+  }
+  return id;
+}
+
+int64 Graph::Count() {
+  // an invariant we mantain is that number of actual nodes == number of nodes in the map
+  return nodemap.size();
+}
+
+void Graph::Connect(int64 from, int64 to) {
+  // check if both nodes are in the graph
+  if(Contains(from) && Contains(to)) {
+    // insert the edge
+    nodemap.at(from).insertOutgoing(to);
+    nodemap.at(to).insertIncoming(from);
+  }
+}
+
+void Graph::Disconnect(int64 from, int64 to) {
+  // check if both nodes are in the graph
+  if(Contains(from) && Contains(to)) {
+    // delete the edge
+    nodemap.at(from).eraseOutgoing(to);
+    nodemap.at(to).eraseIncoming(from);
+  }
+}
+
+bool Graph::IsConnected(int64 from, int64 to) {
+  bool result = false;
+  if(Contains(from)) {
+    result = nodemap.at(from).containsEdgeTo(to);
+  }
+  return result;
+}
+
+Graph Graph::DeepCopy() {
+  // make a new graph and copy over all nodes. The nodes have their own deep
+  // copy method
+  auto result = Graph();
+  for(auto& kv : nodemap) {
+    // k -> id, v -> node
+    result.nodemap.insert({kv.first, kv.second.DeepCopy()});
+  }
+  return result;
+}
+
+void Graph::Reverse(Graph* graph_to_reverse) {
+  // all this does is swap the incoming and outgoing sets for each node
+  auto& nodemap = graph_to_reverse->nodemap;
+  for(auto& kv : nodemap) {
+    // k -> id, v -> node. We don't care about k here...
+    Graph::Node& node = kv.second;
+    std::shared_ptr<std::set<int64> > temp = node.outgoing;
+    node.outgoing = node.incoming;
+    node.incoming = temp;
+  } 
+}
+
+void Graph::Delete(int64 node) {
+  // if the node is in the graph... 
+  if(Contains(node)) {
+    // get the set of nodes which have an edge to it... 
+    std::shared_ptr<std::set<int64> > incoming = nodemap.at(node).incoming;
```

Comment:
You'll have to change this a tiny bit because you're using unique_ptr for the members, but it's simple. Just do the following:

std::set<int64>* incoming = nodemap.at(node).incoming.get() // Note .get()

We use the raw pointer when we want a temporary reference to a unique_ptr. This is nice because it lets you use the thing temporarily without asserting ownership over it, so you don't create any risk of a memory leak.